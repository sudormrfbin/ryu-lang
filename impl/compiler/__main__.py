from compiler.repl import repl

repl()

# with open("input.lang", "r") as f:
#     inp = f.read()

# output = compile(inp)
# print(output)
