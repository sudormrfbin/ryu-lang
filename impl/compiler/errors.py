from typing import ClassVar
import dataclasses
from dataclasses import dataclass
from typing_extensions import override

from lark.tree import Meta
from lark.lexer import Token

from compiler import langtypes

LineCol = tuple[int, int]


@dataclass
class Span:
    start_line: int
    end_line: int
    start_column: int
    end_column: int
    start_pos: int
    end_pos: int

    @classmethod
    def from_meta(cls, meta: Meta) -> "Span":
        return cls(
            start_line=meta.line,
            end_line=meta.end_line,
            start_column=meta.column,
            end_column=meta.end_column,
            start_pos=meta.start_pos,
            end_pos=meta.end_pos,
        )

    @classmethod
    def from_token(cls, token: Token) -> "Span":
        assert token.line is not None
        assert token.end_line is not None
        assert token.column is not None
        assert token.end_column is not None
        assert token.start_pos is not None
        assert token.end_pos is not None

        return cls(
            start_line=token.line,
            end_line=token.end_line,
            start_column=token.column,
            end_column=token.end_column,
            start_pos=token.start_pos,
            end_pos=token.end_pos,
        )

    def coord(self) -> tuple[LineCol, LineCol]:
        start = (self.start_line, self.start_column)
        end = (self.end_line, self.end_column)
        return (start, end)


class InternalCompilerError(Exception):
    pass


@dataclass
class CompilerError(Exception):
    code: ClassVar[int]

    message: str
    span: Span

    def __post_init__(self):
        attributes = dataclasses.astuple(self)
        # Custom exceptions must always call super(), we use the __post_init__
        # mechanism provided by dataclasses to call the super constructor after
        # this class's __init__ is called.
        super().__init__(*attributes)

    @override
    def __str__(self) -> str:
        return f"{type(self).__name__}: {self.message}"


@dataclass
class OperatorSpan:
    name: str
    span: Span


@dataclass
class OperandSpan:
    type_: langtypes.Type
    span: Span


@dataclass
class InvalidOperationError(CompilerError):
    code = 1

    operator: OperatorSpan
    operands: list[OperandSpan]


@dataclass
class UnknownVariable(CompilerError):
    """
    Raised when a variable is used in an expression without declaring it prior.

    ## Example
    ```
    print(var + 5)
    ```
    """

    code = 2

    variable: str


@dataclass
class UndeclaredVariable(CompilerError):
    """
    Raised when the lvalue (the variable part) of an assignment has not
    been declared prior to assignment.

    ## Example
    ```
    var = 4
    ```

    ## Fix
    ```
    let var = 4
    ```
    """

    code = 3

    variable: str


@dataclass
class UnexpectedType(CompilerError):
    """
    Raised when an expression deviates from it's expected type.

    ## Example
    ```
    if 1 {}
    ```

    ## Fix
    ```
    if true {}
    ```
    """

    code = 4

    expected_type: langtypes.Type
    actual_type: langtypes.Type


@dataclass
class TypeMismatch(CompilerError):
    """
    Raised when an expression is expected to be of a particular type
    because of a declaration in another point in the source code.

    ## Example
    ```
    match true {
        case 1 {}
        case 2 {}
    }
    ```

    ## Fix
    ```
    match 1 {
        case 1 {}
        case 2 {}
    }
    ```
    """

    code = 5

    actual_type: langtypes.Type
    expected_type: langtypes.Type
    expected_type_span: Span


@dataclass
class DuplicatedCase(CompilerError):
    """
    Raised when a match statement has duplicated case arms.

    ## Example
    ```
    match true {
        case true {}
        case false {}
    }
    ```

    ## Fix
    ```
    match true {
        case true {}
        case false {}
    }
    ```
    """

    code = 6

    previous_case_span: Span
