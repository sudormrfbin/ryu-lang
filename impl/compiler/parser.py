from typing import Any

from compiler import ast
from compiler.lalr import Token, Tree, Lark_StandAlone, v_args, Transformer, DATA

# https://github.com/lark-parser/lark/issues/565
DATA["options"]["maybe_placeholders"] = True
_parser = Lark_StandAlone(propagate_positions=True)


def parse(source: str) -> Tree[Token]:
    return _parser.parse(source)


def listify(cls):
    @v_args(meta=True, inline=False)
    def _(self, meta, lst):
        return cls(meta, lst)

    return _


@v_args(meta=True, inline=True)
class LarkTreeToAstTransformer(Transformer[Token, Any]):
    def TRUE(self, _) -> bool:
        return True

    def FALSE(self, _) -> bool:
        return False

    def INT(self, n: str) -> int:
        return int(n)

    def STRING(self, s: str) -> str:
        return s[1:-1]  # remove quotes

    statement_list = listify(ast.StatementList)
    statement_block = listify(ast.StatementBlock)
    variable_declaration = ast.VariableDeclaration
    assignment = ast.Assignment
    index_assignment = ast.IndexAssignment
    print_stmt = ast.PrintStmt
    if_chain = ast.IfChain
    if_stmt = ast.IfStmt
    else_if_ladder = listify(ast.ElseIfLadder)
    else_if_stmt = ast.ElseIfStmt
    match_stmt = ast.MatchStmt
    case_ladder = listify(ast.CaseLadder)
    case_stmt = ast.CaseStmt
    enum_pattern = ast.EnumPattern
    enum_pattern_tuple = ast.EnumPatternTuple
    wildcard_pattern = ast.WildcardPattern
    while_stmt = ast.WhileStmt
    for_stmt = ast.ForStmt
    for_stmt_int = ast.ForStmtInt
    struct_stmt = ast.StructStmt
    struct_members = listify(ast.StructMembers)
    struct_member = ast.StructMember
    struct_access = ast.StructAccess
    struct_assignment = ast.StructAssignment
    enum_stmt = ast.EnumStmt
    enum_members = listify(ast.EnumMembers)
    enum_member_bare = ast.EnumMemberBare
    enum_member_tuple = ast.EnumMemberTuple
    function_definition = ast.FunctionDefinition
    function_params = listify(ast.FunctionParams)
    function_param = ast.FunctionParam
    type_annotation = ast.TypeAnnotation
    return_stmt = ast.ReturnStmt
    equality = ast.Equality
    logical = ast.Logical
    comparison = ast.Comparison
    term = ast.Term
    factor = ast.Factor
    unary_op = ast.UnaryOp
    indexing = ast.Indexing
    function_call = ast.FunctionCall
    function_args = listify(ast.FunctionArgs)
    struct_init_members = listify(ast.StructInitMembers)
    struct_init_member = ast.StructInitMember
    bool_literal = ast.BoolLiteral
    int_literal = ast.IntLiteral
    string_literal = ast.StringLiteral
    enum_literal_simple = ast.EnumLiteralSimple
    enum_literal_tuple = ast.EnumLiteralTuple
    variable = ast.Variable
    array_literal = ast.ArrayLiteral
    array_elements = listify(ast.ArrayElements)
    array_element = ast.ArrayElement
    array_pattern = listify(ast.ArrayPattern)
    array_pattern_element = ast.ArrayPatternElement


_transformer = LarkTreeToAstTransformer()


def parse_tree_to_ast(tree: Tree[Token]) -> ast.Ast:
    return _transformer.transform(tree)
