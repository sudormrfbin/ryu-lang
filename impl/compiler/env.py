from typing import Any

from .errors import InternalCompilerError


class RuntimeEnvironment:
    values: dict[str, Any]

    def __init__(self):
        self.values = {}

    def define(self, name: str, value: Any):
        # shadowing is allowed
        self.values[name] = value

    def get(self, name: str) -> Any:
        try:
            return self.values[name]
        except KeyError:
            raise InternalCompilerError(f"Variable {name} not defined at runtime")
