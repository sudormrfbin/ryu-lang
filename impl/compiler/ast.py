import abc
from typing import Any, Optional, Union
import typing
import dataclasses
from dataclasses import dataclass
from typing_extensions import override

from lark import Token, ast_utils
from lark.tree import Meta as LarkMeta

from .env import RuntimeEnvironment, TypeEnvironment

from . import langtypes
from . import errors

_LispAst = list[Union[str, "_LispAst"]]

SKIP_SERIALIZE = "skip_serialize"

# TODO: Narrow down this type
EvalResult = Any

AstDict = dict[typing.Type["_Ast"], dict[str, Any]]

# In AstTypeDict, the key type will always correspond to exactly one value type:
# _Ast -> Type
# str -> AstTypeDict
# Since this cannot be expressed in the type system, we settle for a union type
# which introduces two extra invalid states (_Ast -> AstTypeDict & str -> Type)
AstTypeDict = dict[
    typing.Type["_Ast"] | str, typing.Type[langtypes.Type] | "AstTypeDict"
]


# TODO: explain requirement of underscore by lark
@dataclass
class _Ast(abc.ABC, ast_utils.Ast, ast_utils.WithMeta):
    # InitVar makes meta available on the __post_init__ method
    # and excludes it in the generated __init__.
    meta: dataclasses.InitVar[LarkMeta]
    """Line and column numbers from lark framework.
    Converted to Span for strorage within the class."""

    span: errors.Span = dataclasses.field(init=False, metadata={SKIP_SERIALIZE: True})
    """Line and column number information."""

    # kw_only is required to make dataclasses play nice with inheritance and
    # fields with default values. https://stackoverflow.com/a/69822584/7115678
    type_: langtypes.Type | None = dataclasses.field(
        default=None, kw_only=True, metadata={SKIP_SERIALIZE: True}
    )

    def __post_init__(self, meta: LarkMeta):
        self.span = errors.Span.from_meta(meta)

    @abc.abstractmethod
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        pass

    @abc.abstractmethod
    def eval(self, env: RuntimeEnvironment) -> EvalResult:
        pass

    def to_dict(self) -> AstDict:
        attrs: dict[str, Any] = {}

        for field in dataclasses.fields(self):
            if SKIP_SERIALIZE in field.metadata:
                continue

            value = getattr(self, field.name)
            if isinstance(value, _Ast):
                attrs[field.name] = value.to_dict()
            elif isinstance(value, list) and isinstance(value[0], _Ast):
                attrs[field.name] = [v.to_dict() for v in value]  # type: ignore
            elif value is not None:
                attrs[field.name] = value

        return {type(self): attrs}

    def to_type_dict(
        self,
    ) -> AstTypeDict:
        assert self.type_ is not None

        result: AstTypeDict = {}
        result[type(self)] = type(self.type_)

        for field in dataclasses.fields(self):
            if SKIP_SERIALIZE in field.metadata:
                continue

            value = getattr(self, field.name)
            if isinstance(value, _Ast):
                result[field.name] = value.to_type_dict()
            elif isinstance(value, list) and isinstance(value[0], _Ast):
                result[field.name] = [v.to_type_dict() for v in value]  # type: ignore

        return result


class _Statement(_Ast):
    pass


@dataclass
class StatementList(_Ast, ast_utils.AsList):
    stmts: list[_Statement]

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        for child in self.stmts:
            child.typecheck(env)

        self.type_ = langtypes.BLOCK
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        for child in self.stmts:
            child.eval(env)


@dataclass
class StatementBlock(StatementList):
    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        child_env = TypeEnvironment(enclosing=env)
        return super().typecheck(child_env)

    @override
    def eval(self, env: RuntimeEnvironment):
        child_env = RuntimeEnvironment(enclosing=env)
        return super().eval(child_env)


class _Expression(_Statement):
    pass


@dataclass
class VariableDeclaration(_Statement):
    ident: str
    rvalue: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.type_ = self.rvalue.typecheck(env)
        env.define(self.ident, self.type_)
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        rhs = self.rvalue.eval(env)
        env.define(self.ident, rhs)


@dataclass
class Assignment(_Statement):
    lvalue: Token
    rvalue: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        lvalue_type = env.get(self.lvalue)
        if lvalue_type is None:
            raise errors.UndeclaredVariable(
                message=f"Variable '{self.lvalue}' not declared in this scope",
                span=errors.Span.from_token(self.lvalue),
                variable=self.lvalue,
                # TODO: Add help message to use let
            )

        rvalue_type = self.rvalue.typecheck(env)
        if lvalue_type != rvalue_type:
            raise errors.InternalCompilerError("Type mismatch: TODO")  # TODO

        self.type_ = rvalue_type
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        rhs = self.rvalue.eval(env)
        env.set(self.lvalue, rhs)


@dataclass
class IfStmt(_Statement):
    cond: _Expression
    true_block: StatementBlock

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        expr_type = self.cond.typecheck(env)
        if expr_type != langtypes.BOOL:
            raise errors.UnexpectedType(
                message="Unexpected type for if condition",
                span=self.cond.span,
                expected_type=langtypes.BOOL,
                actual_type=expr_type,
            )

        self.true_block.typecheck(env)

        self.type_ = langtypes.BLOCK
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment) -> bool:
        if self.cond.eval(env) is True:
            self.true_block.eval(env)
            return True

        return False


@dataclass
class IfChain(_Statement):
    if_stmt: IfStmt
    else_if_ladder: Optional["ElseIfLadder"]
    else_block: Optional[StatementBlock]

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.if_stmt.typecheck(env)
        if self.else_block:
            self.else_block.typecheck(env)
        if self.else_if_ladder:
            self.else_if_ladder.typecheck(env)

        self.type_ = langtypes.BLOCK
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        if self.if_stmt.eval(env):
            # if condition evaluates to true, stop the if-chain
            return

        if self.else_if_ladder:
            if self.else_if_ladder.eval(env) is True:
                # one of the else-if blocks executed, stop the if-chain
                return

        if self.else_block:
            self.else_block.eval(env)


@dataclass
class ElseIfStmt(IfStmt):
    pass


@dataclass
class ElseIfLadder(_Statement, ast_utils.AsList):
    blocks: list[ElseIfStmt]

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        for block in self.blocks:
            block.typecheck(env)

        self.type_ = langtypes.BLOCK
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment) -> bool:
        """
        Returns True if any of the else if blocks execute.
        """
        for block in self.blocks:
            if block.eval(env) is True:
                return True
        return False


@dataclass
class CaseStmt(_Ast):
    pattern: "BoolLiteral"
    block: StatementBlock

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.block.typecheck(env)
        self.type_ = self.pattern.typecheck(env)
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment) -> EvalResult:
        self.block.eval(env)


@dataclass
class CaseLadder(_Ast, ast_utils.AsList):
    cases: list[CaseStmt]

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        for case_ in self.cases:
            case_.typecheck(env)

        self.type_ = langtypes.BLOCK
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        pass

    def ensure_exhaustive_matching_bool(self):
        seen: dict[bool, BoolLiteral] = {}
        for case_ in self.cases:
            pattern = case_.pattern.value
            if pattern in seen:
                raise  # TODO
                # raise errors.DuplicatedCase()
            seen[pattern] = case_.pattern

        remaining = {True, False} - set(seen)
        if remaining:
            raise  # TODO
            # raise errors.InexhaustiveMatch()


@dataclass
class MatchStmt(_Statement):
    expr: _Expression
    cases: CaseLadder

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        expr_type = self.expr.typecheck(env)
        self.cases.typecheck(env)

        for case_ in self.cases.cases:
            case_type = case_.pattern.type_
            assert case_type is not None

            if case_type != expr_type:
                # TODO: Add spanshot test when adding more types of patterns
                raise errors.TypeMismatch(
                    message=f"Expected type {expr_type} but got {case_type}",
                    span=case_.pattern.span,
                    actual_type=case_type,
                    expected_type=expr_type,
                    expected_type_span=self.expr.span,
                )

        match expr_type:
            case langtypes.BOOL:
                self.cases.ensure_exhaustive_matching_bool()
            case _:
                raise errors.InternalCompilerError(
                    "TODO: unsupported type for match expression"
                )

        self.type_ = langtypes.BLOCK
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment) -> EvalResult:
        expr = self.expr.eval(env)

        for case_ in self.cases.cases:
            if expr == case_.pattern.value:
                case_.eval(env)
                break
        else:
            raise errors.InternalCompilerError(
                "Match statement did not execute any case blocks"
            )


@dataclass
class Term(_Expression):
    left: _Expression
    op: Token
    right: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        left_type = self.left.typecheck(env)
        right_type = self.right.typecheck(env)

        match left_type, self.op, right_type:
            case langtypes.INT, "+" | "-", langtypes.INT:
                self.type_ = langtypes.INT
            case langtypes.STRING, "+", langtypes.STRING:
                self.type_ = langtypes.STRING
            case _:
                op_span = errors.Span.from_token(self.op)
                raise errors.InvalidOperationError(
                    message=f"Invalid operation {self.op} for types {left_type.name} and {right_type.name}",
                    span=self.span,
                    operator=errors.OperatorSpan(self.op, op_span),
                    operands=[
                        errors.OperandSpan(left_type, self.left.span),
                        errors.OperandSpan(right_type, self.right.span),
                    ],
                )

        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        left = self.left.eval(env)
        right = self.right.eval(env)
        match self.op:
            case "+":
                return left + right
            case "-":
                return left - right
            case _:
                raise errors.InternalCompilerError(
                    f"{type(self).__name__} recieved invalid operator {self.op}"
                )


@dataclass
class Factor(_Expression):
    left: _Expression
    op: Token
    right: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        left_type = self.left.typecheck(env)
        right_type = self.right.typecheck(env)

        match left_type, self.op, right_type:
            case langtypes.INT, "*" | "/" | "%", langtypes.INT:
                self.type_ = langtypes.INT
            case _:
                op_span = errors.Span.from_token(self.op)
                raise errors.InvalidOperationError(
                    message=f"Invalid operation {self.op} for types {left_type.name} and {right_type.name}",
                    span=self.span,
                    operator=errors.OperatorSpan(self.op, op_span),
                    operands=[
                        errors.OperandSpan(left_type, self.left.span),
                        errors.OperandSpan(right_type, self.right.span),
                    ],
                )

        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        left = self.left.eval(env)
        right = self.right.eval(env)
        match self.op:
            case "*":
                return left * right
            case "/":
                match self.left.type_, self.right.type_:
                    case langtypes.INT, langtypes.INT:
                        return left // right
                    case _:
                        raise errors.InternalCompilerError(
                            f"{type(self).__name__} recieved invalid operator {self.op}"
                        )
            case "%":
                match self.left.type_, self.right.type_:
                    case langtypes.INT, langtypes.INT:
                        return left % right
                    case _:
                        raise errors.InternalCompilerError(
                            f"{type(self).__name__} recieved invalid operator {self.op}"
                        )
            case _:
                raise errors.InternalCompilerError(
                    f"{type(self).__name__} recieved invalid operator {self.op}"
                )


@dataclass
class Comparison(_Expression):
    left: _Expression
    op: Token
    right: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        left_type = self.left.typecheck(env)
        right_type = self.right.typecheck(env)

        match left_type, self.op, right_type:
            case langtypes.INT, ">" | "<" | "<=" | ">=", langtypes.INT:
                self.type_ = langtypes.INT
            case _:
                op_span = errors.Span.from_token(self.op)
                raise errors.InvalidOperationError(
                    message=f"Invalid operation {self.op} for types {left_type.name} and {right_type.name}",
                    span=self.span,
                    operator=errors.OperatorSpan(self.op, op_span),
                    operands=[
                        errors.OperandSpan(left_type, self.left.span),
                        errors.OperandSpan(right_type, self.right.span),
                    ],
                )

        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        left = self.left.eval(env)
        right = self.right.eval(env)
        match self.op:
            case ">":
                return left > right
            case "<":
                return left < right
            case "<=":
                return left <= right
            case ">=":
                return left >= right
            case _:
                raise errors.InternalCompilerError(
                    f"{type(self).__name__} recieved invalid operator {self.op}"
                )


@dataclass
class Logical(_Expression):
    left: _Expression
    op: Token
    right: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        left_type = self.left.typecheck(env)
        right_type = self.right.typecheck(env)

        match left_type, self.op, right_type:
            case langtypes.BOOL, "&&", langtypes.BOOL:
                self.type_ = langtypes.BOOL
            case langtypes.BOOL, "||", langtypes.BOOL:
                self.type_ = langtypes.BOOL
            case _:
                op_span = errors.Span.from_token(self.op)
                raise errors.InvalidOperationError(
                    message=f"Invalid operation {self.op} for types {left_type.name} and {right_type.name}",
                    span=self.span,
                    operator=errors.OperatorSpan(self.op, op_span),
                    operands=[
                        errors.OperandSpan(left_type, self.left.span),
                        errors.OperandSpan(right_type, self.right.span),
                    ],
                )

        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        left = self.left.eval(env)
        right = self.right.eval(env)
        match self.op:
            case "&&":
                return left and right
            case "||":
                return left or right
            case _:
                raise errors.InternalCompilerError(
                    f"{type(self).__name__} recieved invalid operator {self.op}"
                )


@dataclass
class Equality(_Expression):
    left: _Expression
    op: Token
    right: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        left_type = self.left.typecheck(env)
        right_type = self.right.typecheck(env)

        match left_type, self.op, right_type:
            case langtypes.INT, "==", langtypes.INT:
                self.type_ = langtypes.BOOL
            case langtypes.INT, "!=", langtypes.INT:
                self.type_ = langtypes.BOOL
            case _:
                op_span = errors.Span.from_token(self.op)
                raise errors.InvalidOperationError(
                    message=f"Invalid operation {self.op} for types {left_type.name} and {right_type.name}",
                    span=self.span,
                    operator=errors.OperatorSpan(self.op, op_span),
                    operands=[
                        errors.OperandSpan(left_type, self.left.span),
                        errors.OperandSpan(right_type, self.right.span),
                    ],
                )

        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        left = self.left.eval(env)
        right = self.right.eval(env)
        match self.op:
            case "==":
                return left == right
            case "!=":
                return left != right
            case _:
                raise errors.InternalCompilerError(
                    f"{type(self).__name__} recieved invalid operator {self.op}"
                )


@dataclass
class UnaryOp(_Expression):
    op: Token
    operand: _Expression

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        operand_type = self.operand.typecheck(env)

        match self.op, operand_type:
            case "+" | "-", langtypes.INT:
                self.type_ = operand_type
            case "!", langtypes.BOOL:
                self.type_ = operand_type
            case _:
                op_span = errors.Span.from_token(self.op)
                raise errors.InvalidOperationError(
                    message=f"Invalid operation '{self.op}' for type '{operand_type.name}'",
                    span=self.span,
                    operator=errors.OperatorSpan(self.op, op_span),
                    operands=[errors.OperandSpan(operand_type, self.operand.span)],
                )

        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        result = self.operand.eval(env)
        match self.op:
            case "+":
                return result
            case "-":
                return -result
            case "!":
                return not result
            case _:
                raise errors.InternalCompilerError(
                    f"{type(self).__name__} recieved invalid operator {self.op}"
                )


@dataclass
class BoolLiteral(_Expression):
    value: bool

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.type_ = langtypes.BOOL
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        return self.value


@dataclass
class IntLiteral(_Expression):
    value: int

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.type_ = langtypes.INT
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        return self.value


@dataclass
class StringLiteral(_Expression):
    value: str

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.type_ = langtypes.STRING
        return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        return self.value


@dataclass
class Variable(_Expression):
    value: str

    @override
    def typecheck(self, env: TypeEnvironment) -> langtypes.Type:
        self.type_ = env.get(self.value)
        if self.type_ is None:
            raise errors.UnknownVariable(
                message=f"Variable '{self.value}' not declared in this scope",
                span=self.span,
                variable=self.value,
            )
        else:
            return self.type_

    @override
    def eval(self, env: RuntimeEnvironment):
        return env.get(self.value)
