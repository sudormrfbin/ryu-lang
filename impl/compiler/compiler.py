from typing import Any

from compiler.env import RuntimeEnvironment, TypeEnvironment

from compiler import builtins, errors, langvalues
from compiler.parser import parse, parse_tree_to_ast

BUILTIN_FUNCTIONS: list[type[langvalues.BuiltinFunction]] = [
    builtins.SumFunction,
    builtins.ArrayLengthFunction,
    builtins.StringLengthFunction,
    builtins.ArrayAppend,
]


def get_default_environs() -> tuple[TypeEnvironment, RuntimeEnvironment]:
    type_env = TypeEnvironment()
    runtime_env = RuntimeEnvironment()

    for fn in BUILTIN_FUNCTIONS:
        type_env.define(fn.TYPE.function_name, fn.TYPE)
        runtime_env.define(fn.TYPE.function_name, fn())

    return (type_env, runtime_env)


def run(source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment) -> Any:
    try:
        return _run(source, type_env, runtime_env)
    except errors.CompilerError as err:
        err.report(source)


def _run(
    source: str, type_env: TypeEnvironment, runtime_env: RuntimeEnvironment
) -> Any:
    tree = parse(source)
    ast = parse_tree_to_ast(tree)

    ast.typecheck(type_env)

    return ast.eval(runtime_env)
