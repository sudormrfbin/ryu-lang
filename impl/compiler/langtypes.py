from __future__ import annotations

from dataclasses import dataclass
from typing import TYPE_CHECKING, Optional, Sequence, TypeAlias
from typing_extensions import override


if TYPE_CHECKING:
    from compiler.env import TypeEnvironment
    from compiler.errors import Span


class Type:
    @property
    def name(self) -> str:
        return type(self).__name__

    @classmethod
    def from_str(cls, ident: str, env: TypeEnvironment) -> Optional["Type"]:
        if ident in PRIMITIVE_TYPES:
            return PRIMITIVE_TYPES[ident]

        ty = env.get(ident)
        if isinstance(ty, UserDefinedType):
            return ty

        return None

    @classmethod
    def from_str_with_generics(
        cls,
        ident: str,
        generics: Type,
        env: TypeEnvironment,
    ) -> Optional["Type"]:
        if ident == "array":
            return Array(generics)

        return None


class Placeholder(Type):
    pass


class PrimitiveType(Type):
    pass


class UserDefinedType(Type):
    pass


class Bool(PrimitiveType):
    pass


class Int(PrimitiveType):
    pass


class String(PrimitiveType):
    pass


class Block(Type):
    pass


@dataclass
class ReturnBlock(Block):
    return_type: Type
    return_stmt_span: Span


@dataclass
class Struct(UserDefinedType):
    struct_name: str
    members: Members

    @dataclass
    class Members(PrimitiveType):
        types: dict[str, Type]

        def __len__(self) -> int:
            return len(self.types)


@dataclass
class Enum(UserDefinedType):
    enum_name: str
    members: list[EnumVariantSimple | EnumVariantTuple]
    span: Span

    @property
    @override
    def name(self) -> str:
        return self.enum_name

    def variant_from_str(
        self, name: str
    ) -> EnumVariantSimple | EnumVariantTuple | None:
        for mem in self.members:
            if mem.name == name:
                return mem
        return None


@dataclass
class EnumVariantSimple:
    name: str


@dataclass
class EnumVariantTuple:
    name: str
    inner: Type


EnumVariants: TypeAlias = EnumVariantSimple | EnumVariantTuple


@dataclass
class Array(PrimitiveType):
    ty: Type

    @property
    @override
    def name(self) -> str:
        return f"Array<{self.ty.name}>"


@dataclass
class UntypedArray(PrimitiveType):
    pass


@dataclass
class Function(PrimitiveType):
    function_name: str
    arguments: Params
    return_type: Type


@dataclass
class Params(PrimitiveType):
    types: list[Type]

    def __len__(self) -> int:
        return len(self.types)


BOOL = Bool()
INT = Int()
STRING = String()
BLOCK = Block()
PLACEHOLDER = Placeholder()


PRIMITIVE_TYPES: dict[str, Type] = {
    "bool": BOOL,
    "int": INT,
    "string": STRING,
}


def resolve_blocks_type(types: Sequence[Type]) -> Block:
    resolved = BLOCK
    for ty in types:
        match (resolved, ty):
            case (Block(), ReturnBlock()):
                resolved = ty
            case (ReturnBlock(), ReturnBlock()):
                if resolved.return_type != ty.return_type:
                    raise  # TODO different return types
            case _:
                pass

    return resolved
