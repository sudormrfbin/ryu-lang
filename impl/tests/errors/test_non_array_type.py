import pytest
from _pytest.capture import CaptureFixture

from compiler.compiler import run
from compiler.env import RuntimeEnvironment, TypeEnvironment
from compiler.errors import UnexpectedType
from compiler.langtypes import Array, INT
from compiler.parser import parse, parse_tree_to_ast

EMPTY_ENV = RuntimeEnvironment()
EMPTY_TYPE_ENV = TypeEnvironment()


def test_for_stmt_non_array_cond():
    with pytest.raises(UnexpectedType) as excinfo:
        ast = parse_tree_to_ast(parse("for i in 43 { print i }"))
        ast.typecheck(EMPTY_TYPE_ENV)

    err = excinfo.value

    assert err.span.coord() == ((1, 10), (1, 12))
    assert err.expected_type == Array(err.actual_type)
    assert err.actual_type == INT


def test_for_stmt_output(capfd: CaptureFixture[str], snapshot: str):
    run("for i in 43 { print i }", EMPTY_TYPE_ENV, EMPTY_ENV)
    _, err = capfd.readouterr()
    assert snapshot == err
