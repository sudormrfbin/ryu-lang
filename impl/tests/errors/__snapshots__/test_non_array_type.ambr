# serializer version: 1
# name: test_for_stmt_output
  '''
  
  [31m[E04] Error:[0m Expected a type of [38;5;201mArray<Int>[0m but found [38;5;155mInt[0m
     [38;5;246mâ•­[0m[38;5;246mâ”€[0m[38;5;246m[[0mstdin:1:10[38;5;246m][0m
     [38;5;246mâ”‚[0m
   [38;5;246m1 â”‚[0m [38;5;249mf[0m[38;5;249mo[0m[38;5;249mr[0m[38;5;249m [0m[38;5;249mi[0m[38;5;249m [0m[38;5;249mi[0m[38;5;249mn[0m[38;5;249m [0m[38;5;155m4[0m[38;5;155m3[0m[38;5;249m [0m[38;5;249m{[0m[38;5;249m [0m[38;5;249mp[0m[38;5;249mr[0m[38;5;249mi[0m[38;5;249mn[0m[38;5;249mt[0m[38;5;249m [0m[38;5;249mi[0m[38;5;249m [0m[38;5;249m}[0m
   [38;5;240m  â”‚[0m          [38;5;155mâ”€[0m[38;5;155mâ”¬[0m  
   [38;5;240m  â”‚[0m           [38;5;155mâ•°[0m[38;5;155mâ”€[0m[38;5;155mâ”€[0m This is of type [38;5;155mInt[0m
  [38;5;246mâ”€â”€â”€â•¯[0m
  
  
  '''
# ---
